File: DAA/DAA cpp/DAA Pract3.cpp
//Fractional knapsack
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

struct Item {
    int weight;
    int value;
    double ratio; // Value per weight ratio
};

// Comparator function to sort items based on value per weight ratio
bool compare(Item a, Item b) {
    return a.ratio > b.ratio;
}

double fractionalKnapsack(int capacity, vector<Item>& items) {
    // Sort items based on value per weight ratio
    sort(items.begin(), items.end(), compare);

    double totalValue = 0.0;

    for (auto& item : items) {
        if (capacity >= item.weight) {
            // If the whole item can be taken
            totalValue += item.value;
            capacity -= item.weight;
        } else {
            // Take the fraction of the item
            double fraction = (double)capacity / item.weight;
            totalValue += item.value * fraction;
            break;
        }
    }

    return totalValue;
}

int main() {
    int numItems, capacity;
    cout << "Enter the number of items: ";
    cin >> numItems;

    vector<Item> items(numItems);

    cout << "Enter the weight and value for each item:\n";
    for (int i = 0; i < numItems; ++i) {
        cout << "Item " << i + 1 << " Weight: ";
        cin >> items[i].weight;
        cout << "Item " << i + 1 << " Value: ";
        cin >> items[i].value;
        items[i].ratio = (double)items[i].value / items[i].weight;
    }

    cout << "Enter the capacity of the knapsack: ";
    cin >> capacity;

    double maxValue = fractionalKnapsack(capacity, items);

    cout << "The maximum value that can be obtained = " << maxValue << endl;

    return 0;
}

File: DAA/DAA python/DAA Pract3.py
def fractional_knapsack(weights, values, capacity):
    n = len(weights)
    # Calculate the value-to-weight ratio for each item
    value_per_weight = [(values[i] / weights[i], weights[i], values[i]) for i in range(n)]
    # Sort items based on the value-to-weight ratio in descending order
    value_per_weight.sort(reverse=True)

    total_value = 0.0
    knapsack = [0.0] * n

    for i in range(n):
        if capacity == 0:
            break

        ratio, weight, value = value_per_weight[i]
        # Take the whole item if it fits
        if weight <= capacity:
            knapsack[i] = 1.0
            total_value += value
            capacity -= weight
        else:
            # Take a fraction of the item
            fraction = capacity / weight
            knapsack[i] = fraction
            total_value += fraction * value
            capacity = 0

    return total_value, knapsack

# Get user input
n = int(input("Enter the number of items: "))
weights = []
values = []

for i in range(n):
    weight = float(input(f"Enter the weight of item {i + 1}: "))
    value = float(input(f"Enter the value of item {i + 1}: "))
    weights.append(weight)
    values.append(value)

capacity = float(input("Enter the capacity of the knapsack: "))

# Calculate and display the result
max_value, knapsack_items = fractional_knapsack(weights, values, capacity)

print("\nMaximum value in the knapsack:", max_value)
print("Fraction of each item in the knapsack:", knapsack_items)
